### PR TITLE
For review

### PR DESCRIPTION
**Pull request by amitkot**
@yaronvel please review

### CODE REVIEW

File: contracts/DxMarketMaker.sol [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+            sellToken,
+            buyToken,
+            auctionIndex,
+            buyTokenAmount
+        );
+        if (!willClearAuction) {
+            buyTokenAmount = addFee(buyTokenAmount);
+        }
+
+        require(
```
Comment (yaronvel) [inline contracts/DxMarketMaker.sol]: I am not sure when do you want to return false and when to revert.

File: contracts/DxMarketMaker.sol [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+            dx.balances(buyToken, address(this)) >= buyTokenAmount,
+            "Not enough buy token to buy required amount"
+        );
+        dx.postBuyOrder(sellToken, buyToken, auctionIndex, buyTokenAmount);
```
Comment (yaronvel) [inline contracts/DxMarketMaker.sol]: is there a return value to check?

File: contracts/DxMarketMaker.sol [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+        emit BoughtInAuction(
+            sellToken,
+            buyToken,
+            auctionIndex,
+            buyTokenAmount,
+            willClearAuction
+        );
+        return true;
```
Comment (yaronvel) [inline contracts/DxMarketMaker.sol]: after reading the code I am sure it can't be a public function.

File: contracts/DxMarketMaker.sol [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+        // uint auctionStart = getAuctionStart(sellToken, buyToken);
+
+        // R2
+        // require(auctionStart <= now);
+
+        // R4
+        // require(auctionIndex == getAuctionIndex(sellToken, buyToken));
+
+        // R5: auction must not be in waiting period
+        // require(auctionStart > AUCTION_START_WAITING_FOR_FUNDING);
```
Comment (yaronvel) [inline contracts/DxMarketMaker.sol]: please remove all these comments if they are no longer relevant

File: contracts/DxMarketMaker.sol [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+        // Overbuy is when a part of a buy order clears an auction
+        // In that case we only process the part before the overbuy
+        // To calculate overbuy, we first get current price
+        uint sellVolume = dx.sellVolumesCurrent(sellToken, buyToken);
+
+        uint num;
+        uint den;
+        (num, den) = dx.getCurrentAuctionPrice(sellToken, buyToken, auctionIndex);
+        // 10^30 * 10^37 = 10^67
+        uint outstandingVolume = atleastZero(int(mul(sellVolume, num) / den - buyVolume));
```
Comment (yaronvel) [inline contracts/DxMarketMaker.sol]: `mul`,`/`,`-`, you are not consistent.

File: contracts/DxMarketMaker.sol [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+    function magic(
+        address sellToken,
+        address buyToken
+    )
+        public
+        returns (bool)
+    {
+        AuctionState state = getAuctionState(sellToken, buyToken);
+
+        if (state == AuctionState.AUCTION_TRIGGERED_WAITING) {
```
Comment (yaronvel) [inline contracts/DxMarketMaker.sol]: is there `switch` in solidity?

File: contracts/DxMarketMaker.sol [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+            return false;
```
Comment (yaronvel) [inline contracts/DxMarketMaker.sol]: what return value suppose to mean?

File: contracts/DxMarketMaker.sol [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+            );
+            uint dNum;
+            uint dDen;
+            (dNum, dDen) = dx.getCurrentAuctionPrice(sellToken, buyToken, auctionIndex);
+            uint kNum;
+            uint kDen;
+            (kNum, kDen) = getKyberRate(sellToken, buyToken, amount);
+
+            // TODO: Check for overflow explicitly?
+            if (mul(dNum, kDen) > mul(kNum, dDen)) {
```
Comment (yaronvel) [inline contracts/DxMarketMaker.sol]: there is a decimal issue when comparing the rate.
The short term solution is to support only 18 decimals token. But please assert it in the code.

File: contracts/DxMarketMaker.sol [hunk 9]
```diff
@@ -0,0 +1,501 @@
...
+                return false;
+            }
+
+            buyInAuction(sellToken, buyToken);
```
Comment (yaronvel) [inline contracts/DxMarketMaker.sol]: return value?

File: contracts/DxMarketMaker.sol [hunk 10]
```diff
@@ -0,0 +1,501 @@
...
+            return true;
+        }
+
+        // Should be unreachable.
+        require(false, "Unknown auction state");
```
Comment (yaronvel) [inline contracts/DxMarketMaker.sol]: `require(false) ==> revert()`.

File: contracts/DxMarketMaker.sol [hunk 11]
```diff
@@ -0,0 +1,501 @@
...
+    // Declared in DutchExchange contract but not public.
+    uint public constant DX_AUCTION_START_WAITING_FOR_FUNDING = 1;
+
+    enum AuctionState {
+        NO_AUCTION_TRIGGERED,
+        AUCTION_TRIGGERED_WAITING,
+        AUCTION_IN_PROGRESS
+    }
+
+    AuctionState constant public NO_AUCTION_TRIGGERED = AuctionState.NO_AUCTION_TRIGGERED;
```
[the remark this answers is not in the document]
Comment (amitkot) [reply contracts/DxMarketMaker.sol]: Yes, added a comment.

File: contracts/DxMarketMaker.sol [hunk 12]
```diff
@@ -0,0 +1,501 @@
+pragma solidity ^0.4.23;
```
[the remark this answers is not in the document]
Comment (amitkot) [reply contracts/DxMarketMaker.sol]: Changed compiler version.
Renamed to KyberDxMarketMaker.

File: contracts/DxMarketMaker.sol [hunk 13]
```diff
@@ -0,0 +1,501 @@
...
+        kyberNetworkProxy = KyberNetworkProxy(_kyberNetworkProxy);
+    }
+
+    // TODO: emit event
+    function depositToDx(address token, uint amount)
+        public
+        onlyAdmin
+        returns (uint)
+    {
+        ERC20(token).approve(dx, amount);
```
[the remark this answers is not in the document]
Comment (amitkot) [reply contracts/DxMarketMaker.sol]: Done.

File: contracts/DxMarketMaker.sol [hunk 14]
```diff
@@ -0,0 +1,501 @@
...
+    // TODO: emit event
+    function withdrawFromDx(address token, uint amount)
+        public
+        onlyAdmin
+        returns (uint)
+    {
+        return dx.withdraw(token, amount);
+    }
+
+    // TODO: consider adding a "safety margin" to compensate for accuracy issues.
```
[the remark this answers is not in the document]
Comment (amitkot) [reply contracts/DxMarketMaker.sol]: Done.

File: contracts/DxMarketMaker.sol [hunk 15]
```diff
@@ -0,0 +1,501 @@
...
+    function thresholdNewAuctionToken(address token)
+        public
+        view
+        returns (uint num)
```
[the remark this answers is not in the document]
Comment (amitkot) [reply contracts/DxMarketMaker.sol]: I think the function name is very clear in this case.
I agree that `num` in the return value may confuse, removed it.

[the remark this answers is not in the document]
Comment (amitkot) [reply contracts/DxMarketMaker.sol]: Removed num.

File: contracts/DxMarketMaker.sol [hunk 16]
```diff
@@ -0,0 +1,501 @@
...
+
+    function addFee(uint amount) public view returns (uint) {
+        uint num;
+        uint den;
+        (num, den) = dx.getFeeRatio(msg.sender);
+
+        // amount / (1 - num / den)
+        return div(
+            mul(amount, den),
+            (den - num)
```
[the remark this answers is not in the document]
Comment (amitkot) [reply contracts/DxMarketMaker.sol]: Switched to safe math here as well.

File: contracts/DxMarketMaker.sol [hunk 17]
```diff
@@ -0,0 +1,501 @@
...
+                return AuctionState.AUCTION_TRIGGERED_WAITING;
+            } else {
+                return AuctionState.AUCTION_IN_PROGRESS;
+            }
+        }
+        return AuctionState.NO_AUCTION_TRIGGERED;
+    }
+
+    // TODO: support token -> token
+    function getKyberRate(address _sellToken, address _buyToken, uint amount)
```
[the remark this answers is not in the document]
Comment (amitkot) [reply contracts/DxMarketMaker.sol]: Removed comment..

File: contracts/DxMarketMaker.sol [hunk 18]
```diff
@@ -0,0 +1,501 @@
...
+        (num, den) = dx.getCurrentAuctionPrice(
+            sellToken,
+            buyToken,
+            auctionIndex
+        );
+
+        // No price for this auction, it is a future one.
+        if (den == 0) return 0;
+
+        return mul(sellVolume, num) / den - buyVolume;
```
[the remark this answers is not in the document]
Comment (amitkot) [reply contracts/DxMarketMaker.sol]: Switched to safe math.

File: contracts/DxMarketMaker.sol [hunk 19]
```diff
@@ -0,0 +1,501 @@
...
+    }
+
+    event ClaimedAuctionTokens(
```
[the remark this answers is not in the document]
Comment (amitkot) [reply contracts/DxMarketMaker.sol]: Renamed. What __is__ the naming best practice these days?

File: contracts/DxMarketMaker.sol [hunk 20]
```diff
@@ -0,0 +1,501 @@
...
+        address sellToken,
+        address buyToken,
+        uint previousLastCompletedAuction,
```
[the remark this answers is not in the document]
Comment (amitkot) [reply contracts/DxMarketMaker.sol]: Indexed all of the addresses, would you recommend indexing the uint fields as well?

File: contracts/DxMarketMaker.sol [hunk 21]
```diff
@@ -0,0 +1,501 @@
...
+        public
+        returns (uint sellerFunds, uint buyerFunds)
+    {
+        uint initialLastClaimed = lastClaimedAuction[sellToken][buyToken];
+
+        uint lastCompletedAuction = dx.getAuctionIndex(sellToken, buyToken) - 1;
+        if (lastCompletedAuction <= initialLastClaimed) return;
+
+        uint amount;
+        for (uint i = lastClaimedAuction[sellToken][buyToken] + 1; i <= lastCompletedAuction; i++) {
```
[the remark this answers is not in the document]
Comment (amitkot) [reply contracts/DxMarketMaker.sol]: Discussing this now, will update with the decision.

File: contracts/DxMarketMaker.sol [hunk 22]
```diff
@@ -0,0 +1,501 @@
...
+        uint buyTokenAmount,
+        bool clearedAuction
+    );
+
+    function buyInAuction(
+        address sellToken,
+        address buyToken
+    )
+        public
+        returns (bool bought)
```
[the remark this answers is not in the document]
Comment (amitkot) [reply contracts/DxMarketMaker.sol]: Changed so that return value name is provided only when there is more than one such value.

**Comment by amitkot** [comment]
@yaronvel Please review updates to contract, tests and JS run script.

File: contracts/KyberDxMarketMaker.sol [hunk 23]
```diff
@@ -479,6 +479,41 @@ contract KyberDxMarketMaker is Withdrawable {
...
+    // FFU
+    function executeTransaction(
+        address destination,
+        uint value,
+        bytes data
+    )
+        public
+        onlyAdmin
+    {
+        if (destination.call.value(value)(data)) {
```
Comment (yaronvel) [inline contracts/KyberDxMarketMaker.sol]: I also told it to ilan, just putting require here is nicer. but nvm.

File: contracts/KyberDxMarketMaker.sol [hunk 24]
```diff
@@ -479,6 +479,41 @@ contract KyberDxMarketMaker is Withdrawable {
...
+
+    function adminBuyInAuction(
+        address sellToken,
+        address buyToken
+    )
+        public
+        onlyAdmin
+        returns (bool bought)
+    {
+        return buyInAuction(sellToken, buyToken);
```
Comment (yaronvel) [inline contracts/KyberDxMarketMaker.sol]: require?

File: contracts/KyberDxMarketMaker.sol (hunk 23 above)
Comment (amitkot) [reply contracts/KyberDxMarketMaker.sol]: Discussed f2f, will consider this change for future version.

File: contracts/KyberDxMarketMaker.sol (hunk 24 above)
Comment (amitkot) [reply contracts/KyberDxMarketMaker.sol]: As discussed, return value is a feature here in addition to the wrapped functionality not putting the contract in an unclear state.

**Pull request merged**